pallets/external-validators/src/lib.rs

```diff
@@ -0,0 +1,697 @@
+            }
+            T::OnEraEnd::on_era_end(active_era.index);
+        }
+
+        /// Plan a new era.
+        ///
+        /// * Bump the current era storage (which holds the latest planned era).
+        /// * Store start session index for the new planned era.
+        /// * Clean old era information.
+        /// * Store staking information for the new planned era
+        ///
+        /// Returns the new validator set.
+        pub fn trigger_new_era(start_session_index: SessionIndex) -> Vec<T::ValidatorId> {
+            // Increment or set current era.
+            let new_planned_era = CurrentEra::<T>::mutate(|s| {
+                *s = Some(s.map(|s| s + 1).unwrap_or(0));
+                s.unwrap()
+            });
+            ErasStartSessionIndex::<T>::insert(&new_planned_era, &start_session_index);
+
+            // Clean old era information.
+            if let Some(old_era) = new_planned_era.checked_sub(T::HistoryDepth::get() + 1) {
+                Self::clear_era_information(old_era);
+            }
+
+            // Returns new validators
+            Self::validators()
+        }
+
+        /// Potentially plan a new era.
+        ///
+        /// Get election result from `T::ElectionProvider`.
+        /// In case election result has more than [`MinimumValidatorCount`] validator trigger a new era.
+        ///
+        /// In case a new era is planned, the new validator set is returned.
+        pub(crate) fn try_trigger_new_era(
+            start_session_index: SessionIndex,
+        ) -> Option<Vec<T::ValidatorId>> {
+            Some(Self::trigger_new_era(start_session_index))
+        }
+
+        /// Clear all era information for given era.
+        pub(crate) fn clear_era_information(era_index: EraIndex) {
+            ErasStartSessionIndex::<T>::remove(era_index);
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_now: BlockNumberFor<T>) -> Weight {
+            // just return the weight of the on_finalize.
+            T::DbWeight::get().reads(1)
+        }
+
+        fn on_finalize(_n: BlockNumberFor<T>) {
+            // Set the start of the first era.
+            if let Some(mut active_era) = <ActiveEra<T>>::get() {
+                if active_era.start.is_none() {
+                    let now_as_millis_u64 = T::UnixTime::now().as_millis().saturated_into::<u64>();
+                    active_era.start = Some(now_as_millis_u64);
+                    // This write only ever happens once, we don't include it in the weight in
+                    // general
+                    ActiveEra::<T>::put(active_era);
+                }
+            }
+            // `on_finalize` weight is tracked in `on_initialize`
+        }
+    }
+}
+
+/// Keeps only the first instance of each element in the input vec. Respects ordering of elements.
+fn remove_duplicates<T: Ord + Clone>(input: Vec<T>) -> Vec<T> {
```

Comment:
There really isnt anything that does this in polkadot-sdk?

Reply:
No

Reply:
ignore